### ISSUE TITLE
Start Circus on ranbomized ports (without hardcoded ports)

### ISSUE THREAD

**Issue by prokher** [circus-tent/circus#1024]
Hi,

I would like to be able to start Circus on randomized ports. I provide `endpoint="tcp://127.0.0.1:0"` and `pubsub_endpoint="tcp://127.0.0.1:0"` which work quite well. Now I want to enable plugin I wrote, but it also started with "...:0":`python -c from circus import plugins;plugins.main() --endpoint tcp://127.0.0.1:0 --pubsub tcp://127.0.0.1:0 --config filename:log.txt mylogger.Logger`.

I suppose plugins must be started with actual port values not with ":0". I could try to fix this, but I would like to hear maintainer opinion before.

What I would like to achieve is to be able to deploy web application without any hardcoded port numbers. I already wrote all the logic connecting my apps (redis, django, ...) in hooks, but I think it would be better to encapsulate it as a plugin (which I will be happy to share with community). Unfortunately I faced this issue running plugin, could you please give me some advice?

**Comment by k4nar** [comment]
Indeed, you cannot use `:0` in your plugin because it expects the actual port. But finding the port once Circus is launched might not be easy.

The easiest solution for you would probably be to use unix sockets instead. As it's a file, you can create it wherever you want, and you control its name. You can do that using `--endpoint ipc://endpoint.sock` for instance.

**Comment by prokher** [comment]
True, but my software needs to work in Windows as well :(

**Comment by k4nar** [comment]
Are you using Circus on Windows right now? If so that's cool, but you might run into troubles :) .

We've made a port some time ago, but I don't think that's it's widely used, and we don't have automatized tests for it :( .

---

Anyway, regarding your issue, I don't really see how we could fix that in Circus.

One way for you would be to add a wrapper script to create the two sockets with `:0`, get the ports, close the sockets and run Circus with those ports.
You might run into a race condition though, if the port is used between your close and the opening by circus, put putting that in a while loop should fix it.

**Comment by prokher** [comment]
About Windows, I am not using it there yet, but I will have to (so you probably get new bugreports soon). Anyway, the fact it is not tested in windows is well, disturbing...

Concerning ":0", my initial idea was the following. Once port is bound process can get its number. Once port is known Circus can use its number when it starts plugins. I understand, that if I start plugins manually I will need some port discover technique, but it is not the case if Circus starts plugins by itself.
